File: Python/3Sum.py
class Solution:
    # @return a list of lists of length 3, [[val1,val2,val3]]
    def threeSum(self, num):
		sortedNum = sorted(num)
		res=[]
		for i in range(len(num)-2):
			a = sortedNum[i]
			j = i+1
			k = len(num)-1
			if i >= 1 and a == sortedNum[i-1]:
				continue
			while j<k:
				b = sortedNum[j]
				c = sortedNum[k]
				if a+b+c==0:
					res.append([a,b,c])
					while j<k:
						j = j + 1
						k = k - 1
						if sortedNum[j]!=b or sortedNum[k]!=c:
							break
				elif a+b+c>0:
					while j<k:
						k = k - 1
						if c != sortedNum[k]:
							break
				else:
					while j<k:
						j = j + 1
						if b != sortedNum[j]:
							break
		return res

File: Java/3Sum.java
class 3Sum{
	public ArrayList<ArrayList<Integer>> threeSum(int[] num) {
		ArrayList<ArrayList<Integer>> res = new ArrayList<ArrayList<Integer>>();
        if(num == null  || num.length == 0)
            return res;
        Arrays.sort(num);
        int i = 0;
        while(i < num.length){
            int j = i + 1;
            int k = num.length - 1;
            if(i > 0 && num[i] == num[i-1]){
                i++;
                continue;
            }
            while(j<k){
            	int a = num[i];
            	int b = num[j];
            	int c = num[k];
            	if(a+b+c == 0){
            		ArrayList<Integer> temp = new ArrayList<Integer>();
            		temp.add(a);
            		temp.add(b);
            		temp.add(c);
            		res.add(temp);
            		while(j < k){
                        j++;
                        k--;
                        if(num[j] != b || num[k] != c)
                            break;
                    }
            	}
            	else if(a+b+c > 0){
            		while(j < k && num[k] == c){
            			k--;
            		}
            	}
            	else{
            		while(j < k && num[j] == b){
            			j++;
            		}
            	}
            }
            i++;
        }
        return res;
	}
}